src/main/kotlin/interactor/GetCityHasCheapestInternetConnectionInteractor.kt

```diff
@@ -8,19 +8,19 @@ class GetCityHasCheapestInternetConnectionInteractor(
 
     fun execute(): CityEntity {
         return dataSource.getAllCitiesData()
-            .filter(::excludeNullSalariesAndNullInternetPriceAndQualityData)
-            .sortedBy(::calculateInternetPercent)
-            .first()
+            .filter(::excludeNullSalariesAndNullInternetPriceAndLowQualityData)
+            .minByOrNull (::calculateInternetPercent)
+            ?: throw IllegalArgumentException("Invalid data")
```

Comment:
`IllegalStateException ` is better in case your function doesn't have any argument.